### PR TITLE
Some build reconfiguration, prepared for other changes in the build system.

### PR DESCRIPTION
**Pull request by ethouris**
This applies the correct file location so that `include` directory now officially disappears. The further fixes for better way to declare proper file lists for headers to be installed and source files - fix to be provided later by @russelltg - this has this feature so far blocked as it was not done portable way.

The stats look much better this time, probably this should fix the problem with zero stats. Note that some of the stats are still zero **and they should better be**.

### CODE REVIEW

**Comment by russelltg** [comment]
As far as the header moves, I usually like to do a file structure like this: 
```
haicrypt/
    include/ << this contains public headers
    src/ << this contains private headers and .cpp files

(and the same for srtcore)
```
Benefits of this design:
1. No need for manifest files etc, just install the whole `include/` directory
2. it's clear which headers belong to which project

If you agree to this change, I'd be willing to file a pull request.

**Comment by rndi** [comment]
Hello @ethouris,

- could you please explain your reasoning for include->common move?
- looks like the tclsh is now a hard dependency for cmake to succeed, not just ./configure? If so, readme update is in order.
- please move stats fixes into separate pull request for the sake of reducing the entropy
- please correct spaces/tabs conventions in your commit, event if they were messed up to begin with
- please consider using git squash to clean up the commits before pushing, that saves time for everyone trying to make sense of the end result of your patch set.

In fact, I would suggest re-submitting this pull request to (1) split it in two to account for stats fixes, and (2) squash all the intermediate steps before we merge history onto master. That is aside from explaining include->common move of course.

Thanks!

**Comment by ethouris** [comment]
Ok, so here's this explained:

1. The `include` directory was a mistake since the beginning (various logics for public headers were in use, this makes things consistent), but it was there partially due to our old legacy. This actually wasn't containing the "public headers" for other projects, but rather header files that should be placed somewhere and no one found a better location for them. This move put things in place, to a subproject where they belong. The specification of public headers is now done in a file named `HEADERS.maf`; all files not mentioned there are private headers (or they are exactly mentioned there as "PRIVATE HEADERS").
   * Just extra explanation for @russelltg : The normal practice is that a public header is being included by the implementation file just for checking things out. So, the possibilities are that a) there's an `-I` option for its path (bad practice), b) the file is included directly as `../include/publicheader.h` (acceptable for "borrowing" headers from a dependent subproject, but for just a public header it looks clumsy), c) the file has a symbolic link in `include` directory (barely portable and causes lots of unexpected problems), d) public headers are in the same directory as sources and private headers, they are just either declared only in installation declaration in the build system (this is practised by Qt library), or they can be specified in a separate file. For me the latter causes least trouble, as long as the build system can make use of `HEADERS.maf` file. I even think this file also should be copied to the header file installation directory to have a clear statement about the protected headers.

2. No, I didn't do this "that hard". This is only a preparation, however this solution is only "prepared", not used (see blocked calls to `GetMafHeaders` macro) - and the sense of it was to provide a living example. This should be rather defined different way, somehow in the CMakeLists.txt file directly. I just wanted to submit this as an example and explanation how things should look like, but didn't have time to provide the proper solution.

3. Ok, just hope that this "stats fix" is the only thing to extract from this bunch

4. Please specify which files you found with this problem. Just one or two names help me at least approach the correct place.

5. If it wasn't for hurrying to prepare the code base for applying other waiting fixes... Ok, I'll see what could be done.

**Comment by rndi** [comment]
Hello @ethouris ,

In the order above:

1. I think I understand. But what stops us from having ./include (or something alike) that has only the public headers and other directories such as ./common with things that are not public and then the build system only exporting things that are in ./include (or something alike)? I think the answer to that would be consistency of the source tree so that you don't have things that are logically connected spread across various directories and possibly to avoid global public headers that declare things that are not logically connected. So overall I agree. Maybe others ( @russelltg ?) want to chime in.

2. Right. But without proper solution we don't have a way for external projects to build against the library. So sounds like we cannot merge this patch set until that solution becomes available?

3. Great thanks!

4. At least a couple of make files I think. If you don't see the problem please do go ahead with the rest of the changes and I will correct before the merge.

5. Thank you! I am sure everyone will benefit in the end :)

-Roman

**Comment by ethouris** [comment]
So, I've created the first portion in this commit; it overwrites the previous one. Not sure if I should create a new pull request and somehow dismiss this one, or just continue with this here.

1. That's it, exactly. The `common` directory is actually for source files that cannot be logically aligned to any of the remaining `haicrypt` and `srtcore`.

2. I needed simply to prepare some initial part because @russelltg has already prepared some concept of greater changes in the build system. I just wanted to put things on the right track and allow him to continue. That's why it's important to push this patch before he could create any greater changes.

4. I saw concerns in CMakeLists.txt file, fixed the tabs to spaces in these places, however the file still has tabs in the language structural statement. I think we need some kinda standard for this and then refactor everything according to it.

**Comment by russelltg** [comment]
Why is it bad practice to have an `-I` flag for source files? If you ask me, it's good practice because it makes the includes look like the includes for people using your repository, which imo is a good thing. Definitely no "right" answer there though, it's a matter of personal preference. 

Also as far as I'm concerned, my pull request doesn't rely on yours. How does this enhance/enable my changes?

File: common/appcommon.hpp [hunk 1]
```diff
@@ -37,12 +37,24 @@
 #else
 extern "C" inline int sleep(int seconds) { Sleep(seconds * 1000); return 0; }
 #endif
+
+inline bool SysInitializeNetwork()
+{
+    WORD wVersionRequested = MAKEWORD(2, 2);
+    WSADATA wsaData;
+    return WSAStartup(wVersionRequested, &wsaData) == 0;
```
Comment (russelltg) [inline common/appcommon.hpp]: This is called in `srt_startup`, any reason  you aren't using that instead?

**Comment by russelltg** [comment]
Also it seems like that last commit doesn't really relate to this pull request, maybe it was meant for a different branch? Also seems to be addressed in #37.

File: common/appcommon.hpp (hunk 1 above)
Comment (ethouris) [inline common/appcommon.hpp]: This *could* be fixed by calling `srt_startup()`, but this is logically improper. This is required not for SRT (SRT does it automatically by calling `srt_startup()` together with the first socket), but for UDP, or simply any kind of system networking functions on Windows. In other words, if `stransmit` didn't use any network functions except those from SRT, this wouldn't be necessary at all.

But calling `srt_startup()` just in order to make UDP-related system function work is, you know, kinda clumsy.

**Comment by ethouris** [comment]
I just started with github and I may make mistakes. The last commit was **NOT** intended to be attached to this pull request. I have created a pull request and then I have pushed another commit **to my own branch**, just in order to create y'another pull request out of it in future. I'll try to straighten things up...

**Comment by ethouris** [comment]
Look ma! The extra commit disappeared. Ok, so please check this in, if there are no more concerns.

**Comment by rndi** [comment]
Hello @ethouris ,

That looks good, thank you!

File: common/appcommon.hpp (hunk 1 above)
Comment (russelltg) [inline common/appcommon.hpp]: That makes lots of sense 👍

**Pull request merged**